### PR TITLE
Docs

### PR DESCRIPTION
**Pull request by Legacy107**
- Add docs for code flow
- Add docs for GameEvent, NodeUtil, MathUtil and MapInfo

### CODE REVIEW

File: docs/authors.rst [hunk 1]
```diff
@@ -3,5 +3,6 @@ Authors
 
 * Legacy107
 * sangpham2710
```
Comment (LeStolz) [inline docs/authors.rst]: You can remove "* sangpham2710"

File: docs/codeDocumentation.rst [hunk 2]
```diff
@@ -0,0 +1,10 @@
+Code documentation
+==================
+
+.. toctree::
+    :maxdepth: 1
+    :caption: Contents:
+    :glob:
+
+    flow*
```
Comment (LeStolz) [inline docs/codeDocumentation.rst]: flowS?

File: docs/flow.rst [hunk 3]
```diff
@@ -0,0 +1,9 @@
+Code Flow
+=========
+
+Game Flow
+---------
+
+.. image:: .\\images\\gameFLow.jpg
```
Comment (LeStolz) [inline docs/flow.rst]: "gameFLow" should be "gameFlow"

File: docs/guidelines.rst [hunk 4]
```diff
@@ -73,14 +85,20 @@ Code structure
...
     scenes : folder
-        Containing all primary scenes such as HUD and match maps.
+        Containing all primary scenes such as the HUD and match maps.
 
     scripts : folder
-        Containing scritps controlling player, abilities, game logic, networking, etc.
+        Containing scripts controlling player, abilities, game logics, networking, etc.
+
+    singletons : folder
+        Containing singletons scripts
```
Comment (LeStolz) [inline docs/guidelines.rst]: "singletons" should be "singleton"

File: docs/guidelines.rst [hunk 5]
```diff
@@ -51,19 +60,22 @@ Code structure
...
     
     sfx : folder
-        Containing sounds and musics file.
+        Containing sound and music files.
 
     sprites : folder
         Containing game sprites.
 
+    tileset : folder
+        Containing files relating to tileset such as .tres files.
```
Comment (LeStolz) [inline docs/guidelines.rst]: related?

File: docs/references.rst [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
++====================+=========+
+| CTF_flag_positions |         |
++--------------------+---------+
+
+
+.. topic:: Attributes
+
+    .. glossary::
+        CTF_flag_positions
+            [CTF] The postions of flag stations.
```
Comment (LeStolz) [inline docs/references.rst]: positions

File: docs/references.rst [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+|| :term:`CTF_return_flag`  ||        |
+|| :term:`increase_score`   ||        |
++---------------------------+---------+
+
+
+.. topic:: Signals
+
+    .. glossary::
+        CTF_capture_flag
+            [CTF] Emitted when a team capture the enemy's flag.
```
Comment (LeStolz) [inline docs/references.rst]: captureS

File: docs/references.rst [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+
+        CTF_return_flag
+            [CTF] Emitted when a flag is returned to its station.
+
+            **Parameters**
+
+            - team_id (`int <https://docs.godotengine.org/en/stable/getting_started/scripting/gdscript/gdscript_basics.html#int>`_)
+
+        increase_score
+            Emitted when a team score points.
```
Comment (LeStolz) [inline docs/references.rst]: scoreS

File: docs/guidelines.rst [hunk 9]
```diff
@@ -16,33 +16,42 @@ Code structure
...
     ├─── :term:`assets`
     │   ├─── :term:`images`
     │   ├─── :term:`sfx`
-    │   └─── :term:`sprites`
+    │   ├─── :term:`sprites`
+    │   └─── :term:`tileset`
+    |
     ├─── :term:`docs`
     └─── :term:`src`
         │    :term:`main.tscn`
```
Comment (LeStolz) [inline docs/guidelines.rst]: Do we even have "main.tscn"?

File: docs/guidelines.rst [hunk 10]
```diff
@@ -51,19 +60,22 @@ Code structure
         Containing assets including images, sprites, sfx.
```
Comment (LeStolz) [inline docs/guidelines.rst]: Change "including" to "such as" or "that are"

File: docs/guidelines.rst [hunk 11]
```diff
@@ -51,19 +60,22 @@ Code structure
...
 
+    tileset : folder
+        Containing files relating to tileset such as .tres files.
+
     docs : folder
         Containing documentation files.
     
     src : folder
-        Containing soucrce code including scripts and scenes.
+        Containing source code including scripts and scenes.
```
Comment (LeStolz) [inline docs/guidelines.rst]: Change "including" to "such as" or "that are"

Comment (LeStolz) [inline docs/guidelines.rst]: Shouldn't it be "source codes"?

File: docs/guidelines.rst [hunk 12]
```diff
@@ -73,14 +85,20 @@ Code structure
         to abilities, particles, etc.
 
     scenes : folder
-        Containing all primary scenes such as HUD and match maps.
+        Containing all primary scenes such as the HUD and match maps.
 
     scripts : folder
-        Containing scritps controlling player, abilities, game logic, networking, etc.
+        Containing scripts controlling player, abilities, game logics, networking, etc.
```
Comment (LeStolz) [inline docs/guidelines.rst]: Maybe change "controlling" to "that control"

Comment (LeStolz) [inline docs/guidelines.rst]: "player" should be "players"

File: docs/references.rst [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+    .. glossary::
+        CTF_capture_flag
+            [CTF] Emitted when a team capture the enemy's flag.
+
+            **Parameters**
+
+            - team_id (`int <https://docs.godotengine.org/en/stable/getting_started/scripting/gdscript/gdscript_basics.html#int>`_)
+
+        CTF_return_flag
+            [CTF] Emitted when a flag is returned to its station.
```
Comment (LeStolz) [inline docs/references.rst]: Maybe you should say "to its team's station" to make it more obvious what the team_id param is for

File: docs/references.rst (hunk 8 above)
Comment (LeStolz) [inline docs/references.rst]: "Emitted when a team scores." is good enough and is better in that it make the score param more obvious

File: docs/references.rst [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+    .. glossary::
+        calculate_pivot_rotation
+            Calculates the Pivot rotation such that Origin's look vector points at Target's global_position.
+
+            **Parameters**
+
+            - Pivot (`Position2D <https://docs.godotengine.org/en/stable/classes/class_position2d.html>`_) –
+              The pivot and the object (as a child of the pivot) whose rotation is calculated.
+            - Target (`Position2D <https://docs.godotengine.org/en/stable/classes/class_position2d.html>`_) –
+              The target that the Origin wants to point at.
```
Comment (LeStolz) [inline docs/references.rst]: "Origin" should be "Origin's look vector"

File: docs/references.rst [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+            - Origin (`Position2D <https://docs.godotengine.org/en/stable/classes/class_position2d.html>`_) –
+              The origin of the object whose rotation is calculated.
+
```
Comment (LeStolz) [inline docs/references.rst:132]: Technically the Parameters above can be anything as long as they have position and global_position

File: docs/references.rst [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+        calculate_pivot_rotation
+            Calculates the Pivot rotation such that Origin's look vector points at Target's global_position.
+
+            **Parameters**
+
+            - Pivot (`Position2D <https://docs.godotengine.org/en/stable/classes/class_position2d.html>`_) –
+              The pivot and the object (as a child of the pivot) whose rotation is calculated.
+            - Target (`Position2D <https://docs.godotengine.org/en/stable/classes/class_position2d.html>`_) –
+              The target that the Origin wants to point at.
+            - Origin (`Position2D <https://docs.godotengine.org/en/stable/classes/class_position2d.html>`_) –
```
Comment (LeStolz) [inline docs/references.rst]: Maybe note that the Origin MUST be a direct child of Pivot

File: docs/references.rst [hunk 17]
```diff
@@ -0,0 +1,166 @@
...
+              The origin of the object whose rotation is calculated.
+
+            **Returns**
+
+            The Pivot's rotation angle in radian.
```
Comment (LeStolz) [inline docs/references.rst]: I think you can omit the "angle" ._. and "rotation" is in radian by default so maybe dun need to specify that?

**Review by LeStolz — CHANGES_REQUESTED** [review]
"gameFLow.jpg" should be "gameFlow.jpg" and may I ask why .jpg?

**Comment by Legacy107** [comment]
> "gameFLow.jpg" should be "gameFlow.jpg" and may I ask why .jpg?

JPEG format offers smaller file size

File: docs/guidelines.rst (hunk 9 above)
Comment (Legacy107) [reply docs/guidelines.rst]: We will

File: docs/references.rst [hunk 18]
```diff
@@ -0,0 +1,170 @@
...
+    .. glossary::
+        calculate_pivot_rotation
+            Calculates the Pivot rotation such that Origin's look vector points at Target's global_position.
+
+            **Parameters**
+
+            - Pivot (inherits `Node2D <https://docs.godotengine.org/en/stable/classes/class_node2d.html>`_) –
+              The pivot and the object (as a child of the pivot) whose rotation is calculated.
+            - Target (inherits `Node2D <https://docs.godotengine.org/en/stable/classes/class_node2d.html>`_) –
+              The target that the Origin's look vector wants to point at.
```
Comment (LeStolz) [inline docs/references.rst]: "the Origin" should be "Origin"

File: docs/references.rst [hunk 19]
```diff
@@ -0,0 +1,170 @@
...
+            - Origin (inherits `Node2D <https://docs.godotengine.org/en/stable/classes/class_node2d.html>`_) –
+              The origin of the object whose rotation is calculated.
+
+            .. note::
+
+              The Origin **must** be a direct child of the Pivot.
```
Comment (LeStolz) [inline docs/references.rst]: "The Origin" should be "Origin"

File: docs/guidelines.rst [hunk 20]
```diff
@@ -16,54 +16,66 @@ Code structure
...
     ├─── :term:`assets`
     │   ├─── :term:`images`
     │   ├─── :term:`sfx`
-    │   └─── :term:`sprites`
+    │   ├─── :term:`sprites`
+    │   └─── :term:`tileset`
+    |
     ├─── :term:`docs`
     └─── :term:`src`
         │    :term:`main.tscn`
```
Comment (LeStolz) [inline docs/guidelines.rst]: "main" should be "Main"

File: docs/guidelines.rst [hunk 21]
```diff
@@ -16,54 +16,66 @@ Code structure
...
+        Containing files related to tileset such as .tres files.
+
     docs : folder
         Containing documentation files.
     
     src : folder
-        Containing soucrce code including scripts and scenes.
+        Containing source codes such as scripts and scenes.
 
     main.tscn : file
```
Comment (LeStolz) [inline docs/guidelines.rst]: "main" should be "Main"

**Review by LeStolz — CHANGES_REQUESTED** [review]
Change the goddamn "gameFLow" name properly

File: docs/references.rst [hunk 22]
```diff
@@ -0,0 +1,170 @@
...
++============+==================================+
+|            | :term:`calculate_pivot_rotation` |
++------------+----------------------------------+
+
+
+.. topic:: Methods
+
+    .. glossary::
+        calculate_pivot_rotation
+            Calculates the Pivot rotation such that Origin's look vector points at Target's global_position.
```
Comment (LeStolz) [inline docs/references.rst]: "the Pivot rotation" should be "the Pivot's rotation"

**Review by LeStolz — APPROVED** [review]

**Comment by LeStolz** [comment]
LGTM!

**Pull request merged**